File: payola/web/public/javascripts/cz/payola/web/client/views/graph/GraphView.js
goog.provide('cz.payola.web.client.views.graph.GraphView');
goog.require('cz.payola.web.client.views.Constants');
goog.require('cz.payola.web.client.views.Layer');
goog.require('cz.payola.web.client.views.Point');
goog.require('cz.payola.web.client.views.Vector');
goog.require('cz.payola.web.client.views.graph.Controls');
goog.require('cz.payola.web.client.views.graph.EdgeView');
goog.require('cz.payola.web.client.views.graph.LocationDescriptor');
goog.require('cz.payola.web.client.views.graph.RedrawOperation');
goog.require('cz.payola.web.client.views.graph.VertexView');
goog.require('cz.payola.web.client.views.graph.View');
goog.require('scala.collection.immutable.List');
goog.require('scala.collection.mutable.ListBuffer');
goog.require('scala.math');
cz.payola.web.client.views.graph.GraphView = function(graphModel, container) {
var self = this;
self.graphModel = graphModel;
self.container = container;
self.edgesDeselectedLayer = self.createLayer(self.container);
self.edgesDeselectedTextLayer = self.createLayer(self.container);
self.edgesSelectedLayer = self.createLayer(self.container);
self.edgesSelectedTextLayer = self.createLayer(self.container);
self.verticesDeselectedLayer = self.createLayer(self.container);
self.verticesDeselectedTextLayer = self.createLayer(self.container);
self.verticesSelectedLayer = self.createLayer(self.container);
self.verticesSelectedTextLayer = self.createLayer(self.container);
self.topBlankLayer = self.createLayer(self.container);
self.layers = scala.collection.immutable.List.$apply(self.edgesDeselectedLayer, self.edgesDeselectedTextLayer, self.edgesSelectedLayer, self.edgesSelectedTextLayer, self.verticesDeselectedLayer, self.verticesDeselectedTextLayer, self.verticesSelectedLayer, self.verticesSelectedTextLayer, self.topBlankLayer);
self.controlsLayer = new cz.payola.web.client.views.graph.Controls(self, self.topBlankLayer);
self.vertexViews = self.createVertexViews();
self.edgeViews = self.createEdges();
goog.base(self);};
goog.inherits(cz.payola.web.client.views.graph.GraphView, cz.payola.web.client.views.graph.View);
cz.payola.web.client.views.graph.GraphView.prototype.initControls = function() {
var self = this;
self.controlsLayer.init();
};
cz.payola.web.client.views.graph.GraphView.prototype.createVertexViews = function() {
var self = this;
var graphWidth = 500;
var buffer = scala.collection.mutable.ListBuffer.$apply();
var counter = 0;
var spaceBetweenVertices = 80;
self.graphModel.vertices.foreach(function(vertexModel) { var point = new cz.payola.web.client.views.Point((((counter * spaceBetweenVertices) % graphWidth) + 20), ((scala.math.floor(((counter * spaceBetweenVertices) / graphWidth)) * spaceBetweenVertices) + 20));
buffer.$plus$eq(new cz.payola.web.client.views.graph.VertexView(vertexModel, point));
counter = (counter + 1);
 });
return buffer;
};
cz.payola.web.client.views.graph.GraphView.prototype.createEdges = function() {
var self = this;
var buffer = scala.collection.mutable.ListBuffer.$apply();
self.graphModel.edges.foreach(function(edgeModel) { return buffer.$plus$eq(new cz.payola.web.client.views.graph.EdgeView(edgeModel, self.findVertexView(edgeModel.origin), self.findVertexView(edgeModel.destination)));
 });
return buffer;
};
cz.payola.web.client.views.graph.GraphView.prototype.findVertexView = function(vertexModel) {
var self = this;
return self.vertexViews.find(function($x$1) { return ($x$1.vertexModel === vertexModel);
 }).get();
};
cz.payola.web.client.views.graph.GraphView.prototype.draw = function(context, color, position) {
var self = this;
var positionCorrection = (function() {
if ((position != null)) {
return position.toVector();
} else {
return cz.payola.web.client.views.Point.Zero.toVector();
}})();
self.vertexViews.foreach(function(vertexView) { var colorToUse = (function() {
if ((color != null)) {
return color;
} else {
return (function() {
if (vertexView.selected) {
return cz.payola.web.client.views.Constants.ColorVertexHigh;
} else {
return (function() {
if (self.edgeViews.exists(function(edgeView) { return (((edgeView.originView === vertexView) && edgeView.destinationView.selected) || ((edgeView.destinationView === vertexView) && edgeView.originView.selected));
 })) {
return cz.payola.web.client.views.Constants.ColorVertexMedium;
} else {
return (function() {
if ((self.controlsLayer.selectedCount == 0)) {
return cz.payola.web.client.views.Constants.ColorVertexDefault;
} else {
return cz.payola.web.client.views.Constants.ColorVertexLow;
}})();
}})();
}})();
}})();
(function() {
if (vertexView.selected) {
(function() {
if (self.verticesSelectedLayer.cleared) {
vertexView.draw(self.verticesSelectedLayer.context, colorToUse, position);
} else {
}})();
(function() {
if (self.verticesSelectedTextLayer.cleared) {
vertexView.information.draw(self.verticesSelectedTextLayer.context, color, cz.payola.web.client.views.graph.LocationDescriptor.getVertexInformationPosition(vertexView.position).$plus(positionCorrection));
} else {
}})();
} else {
(function() {
if (self.verticesDeselectedLayer.cleared) {
vertexView.draw(self.verticesDeselectedLayer.context, colorToUse, position);
} else {
}})();
(function() {
if (self.verticesDeselectedTextLayer.cleared) {
vertexView.information.draw(self.verticesDeselectedTextLayer.context, color, cz.payola.web.client.views.graph.LocationDescriptor.getVertexInformationPosition(vertexView.position).$plus(positionCorrection));
} else {
}})();
}})();
 });
self.edgeViews.foreach(function(edgeView) { var positionToUse = cz.payola.web.client.views.graph.LocationDescriptor.getEdgeInformationPosition(edgeView.originView.position, edgeView.destinationView.position).$plus(positionCorrection);
var colorToUse = (function() {
if ((color != null)) {
return color;
} else {
return (function() {
if (edgeView.isSelected()) {
return cz.payola.web.client.views.Constants.ColorEdgeSelect;
} else {
return cz.payola.web.client.views.Constants.ColorEdge;
}})();
}})();
(function() {
if (edgeView.isSelected()) {
(function() {
if (self.edgesSelectedLayer.cleared) {
edgeView.draw(self.edgesSelectedLayer.context, color, position);
} else {
}})();
(function() {
if (self.edgesSelectedTextLayer.cleared) {
edgeView.information.draw(self.edgesSelectedTextLayer.context, colorToUse, positionToUse);
} else {
}})();
} else {
(function() {
if (self.edgesDeselectedLayer.cleared) {
edgeView.draw(self.edgesDeselectedLayer.context, color, position);
} else {
}})();
(function() {
if (self.edgesDeselectedTextLayer.cleared) {
edgeView.information.draw(self.edgesDeselectedTextLayer.context, colorToUse, positionToUse);
} else {
}})();
}})();
 });
self.layers.foreach(function(layer) { layer.cleared = false;
 });
};
cz.payola.web.client.views.graph.GraphView.prototype.redraw = function(graphOperation) {
var self = this;
(function($selector_1) {
if ($selector_1 === cz.payola.web.client.views.graph.RedrawOperation.Movement) {
self.clear(self.edgesSelectedLayer.context, cz.payola.web.client.views.Point.Zero, self.edgesSelectedLayer.getSize());
self.edgesSelectedLayer.cleared = true;
self.clear(self.edgesSelectedTextLayer.context, cz.payola.web.client.views.Point.Zero, self.edgesSelectedTextLayer.getSize());
self.edgesSelectedTextLayer.cleared = true;
self.clear(self.verticesSelectedLayer.context, cz.payola.web.client.views.Point.Zero, self.verticesSelectedLayer.getSize());
self.verticesSelectedLayer.cleared = true;
self.clear(self.verticesSelectedTextLayer.context, cz.payola.web.client.views.Point.Zero, self.verticesSelectedTextLayer.getSize());
self.verticesSelectedTextLayer.cleared = true;
self.draw(null, null, null);
return;
}
if ($selector_1 === cz.payola.web.client.views.graph.RedrawOperation.Selection) {
self.redrawAll();
return;
}
if (true) {
self.redrawAll();
return;
}
})(graphOperation);
};
cz.payola.web.client.views.graph.GraphView.prototype.redrawAll = function() {
var self = this;
self.layers.foreach(function(layer) { self.clear(layer.context, cz.payola.web.client.views.Point.Zero, layer.getSize());
layer.cleared = true;
 });
self.draw(null, null, null);
};
cz.payola.web.client.views.graph.GraphView.prototype.clear = function(context, topLeft, size) {
var self = this;
var bottomRight = topLeft.$plus(size);
context.clearRect(topLeft.x, topLeft.y, bottomRight.x, bottomRight.y);
};
cz.payola.web.client.views.graph.GraphView.prototype.createLayer = function(container) {
var self = this;
var canvas = document.createElement('canvas');
var context = canvas.getContext('2d');
var layer = new cz.payola.web.client.views.Layer(canvas, context);
container.appendChild(canvas);
layer.setSize(new cz.payola.web.client.views.Vector(600.0, 500.0));
return layer;
};
cz.payola.web.client.views.graph.GraphView.prototype.metaClass_ = new s2js.MetaClass('cz.payola.web.client.views.graph.GraphView', [cz.payola.web.client.views.graph.View]);

File: payola/web/client/src/main/scala/cz/payola/web/client/views/graph/GraphView.scala
package cz.payola.web.client.views.graph

import algorithms.circlePath.CirclePathModel
import algorithms.gravity.GravityModel
import algorithms.pathLength.PathLengthModel
import cz.payola.web.client.views.Constants._
import collection.mutable.ListBuffer
import s2js.adapters.js.dom.{Element, CanvasRenderingContext2D}
import cz.payola.web.client.views.{Color, Point}
import cz.payola.common.rdf.{Vertex, Graph}

class GraphView(val graphModel: Graph, val container: Element) extends View
{
    /*The order in which are layers created determines their "z coordinate"
(first created layer is on the bottom and last created one covers all the others).*/
    private val edgesDeselectedLayer = createLayer(container)

    private val edgesDeselectedTextLayer = createLayer(container)

    private val edgesSelectedLayer = createLayer(container)

    private val edgesSelectedTextLayer = createLayer(container)

    private val verticesDeselectedLayer = createLayer(container)

    private val verticesDeselectedTextLayer = createLayer(container)

    private val verticesSelectedLayer = createLayer(container)

    private val verticesSelectedTextLayer = createLayer(container)

    private val topBlankLayer = createLayer(container)

    private val layers = List(
        edgesDeselectedLayer, edgesDeselectedTextLayer, edgesSelectedLayer, edgesSelectedTextLayer,
        verticesDeselectedLayer, verticesDeselectedTextLayer, verticesSelectedLayer, verticesSelectedTextLayer,
        topBlankLayer)

    private val controlsLayer = new Controls(this, topBlankLayer)

    val vertexViews = createVertexViews()

    val edgeViews = createEdges()

    def init() {
        controlsLayer.init()
        val model = new CirclePathModel()
        model.perform(vertexViews, edgeViews)
    }

    /**
      * Constructs a list of vertexViews based on the graphModel parameter.
      */
    def createVertexViews(): ListBuffer[VertexView] = {
        val buffer = ListBuffer[VertexView]()
        var counter = 0

        graphModel.vertices.foreach {vertexModel =>

            buffer += new VertexView(vertexModel, Point(0, 0))
            counter += 1
        }
        buffer
    }

    /**
      * Constructs a list of edgeViews based on the graphModel and verticesView parameters.
      */
    def createEdges(): ListBuffer[EdgeView] = {
        val buffer = ListBuffer[EdgeView]()
        graphModel.edges.foreach {edgeModel =>
            buffer += new EdgeView(edgeModel, findVertexView(edgeModel.origin), findVertexView(edgeModel.destination))
        }

        vertexViews.foreach {vertexView: VertexView =>
            vertexView.edges = getEdgesOfVertex(vertexView, buffer)
        }

        buffer
    }

    def getEdgesOfVertex(vertexView: VertexView, _edgeViews: ListBuffer[EdgeView]): ListBuffer[EdgeView] = {
        var edgeViewsBuffer = ListBuffer[EdgeView]()
        var i = 0
        _edgeViews.foreach {_edgeView: EdgeView =>
            if ((_edgeView.originView.vertexModel eq vertexView.vertexModel) ||
                (_edgeView.destinationView.vertexModel eq vertexView.vertexModel)) {

                edgeViewsBuffer += _edgeView
                i += 1
            }
        }
        edgeViewsBuffer
    }

    private def findVertexView(vertexModel: Vertex): VertexView = {
        vertexViews.find(_.vertexModel.eq(vertexModel)).get
    }

    def draw(context: CanvasRenderingContext2D, color: Option[Color], position: Option[Point]) {
        val positionCorrection = position match {
            case None => Point.Zero
            case _ => position.get
        }

        vertexViews.foreach {vertexView =>

            var colorToUseVertex = Color.Black
            var colorToUseText = Color.Black
            if (color != None) {
                colorToUseVertex = color.get
                colorToUseText = color.get
            } else if (vertexView.selected) {
                colorToUseVertex = ColorVertexHigh
                colorToUseText = ColorTextHigh
            } else if (edgeViews.exists(edgeView =>
                (edgeView.originView.eq(vertexView) && edgeView.destinationView.selected) ||
                    (edgeView.destinationView.eq(vertexView) && edgeView.originView.selected))) {
                colorToUseVertex = ColorVertexMedium
                colorToUseText = ColorTextHigh
            } else if (controlsLayer.selectedCount == 0) {
                colorToUseVertex = ColorVertexDefault
                colorToUseText = ColorTextHigh
            } else {
                colorToUseVertex = ColorVertexLow
                colorToUseText = Color.Transparent
            }

            if (vertexView.selected) {
                if (verticesSelectedLayer.cleared) {
                    vertexView.draw(verticesSelectedLayer.context, Some(colorToUseVertex), Some(positionCorrection))
                }
                if (verticesSelectedTextLayer.cleared && vertexView.information != None) {
                    vertexView.information.get.draw(verticesSelectedTextLayer.context, Some(colorToUseText),
                        Some(LocationDescriptor.getVertexInformationPosition(vertexView.position) +
                            positionCorrection.toVector))
                }
            } else {
                if (verticesDeselectedLayer.cleared) {
                    vertexView.draw(verticesDeselectedLayer.context, Some(colorToUseVertex), Some(positionCorrection))
                }
                if (verticesDeselectedTextLayer.cleared && vertexView.information != None) {
                    vertexView.information.get.draw(verticesDeselectedTextLayer.context, Some(colorToUseText),
                        Some(LocationDescriptor.getVertexInformationPosition(vertexView.position) +
                            positionCorrection.toVector))
                }
            }
        }

        edgeViews.foreach {edgeView =>

            val positionToUse = LocationDescriptor.getEdgeInformationPosition(
                edgeView.originView.position, edgeView.destinationView.position) + positionCorrection.toVector
            val colorToUseEdge = if (color != None) {
                color.get
            } else if (edgeView.isSelected) {
                ColorEdgeSelect
            } else {
                ColorEdge
            }
            val colorToUseText = if (color != None) {
                color.get
            } else if (edgeView.isSelected) {
                ColorTextHigh
            } else {
                ColorText
            }


            if(edgeView.isSelected) {
                edgeView.information.setSelectedForDrawing()
            }

            if (edgeView.isSelected) {
                if (edgesSelectedLayer.cleared) {
                    edgeView.draw(edgesSelectedLayer.context, Some(colorToUseEdge), Some(positionCorrection))
                }
                if (edgesSelectedTextLayer.cleared) {
                    edgeView.information.draw(edgesSelectedTextLayer.context, Some(colorToUseText), Some(positionToUse))
                }
            } else {
                if (edgesDeselectedLayer.cleared) {
                    edgeView.draw(edgesDeselectedLayer.context, Some(colorToUseEdge), Some(positionCorrection))
                }
                if (edgesDeselectedTextLayer.cleared) {
                    edgeView.information.draw(edgesDeselectedTextLayer.context, Some(colorToUseText), Some(positionToUse))
                }
            }
        }

        layers.foreach(layer => layer.cleared = false)
    }

    def redraw(graphOperation: Int) {
        graphOperation match {
            case RedrawOperation.Movement =>
                clear(edgesSelectedLayer.context, Point.Zero, edgesSelectedLayer.getSize)
                edgesSelectedLayer.cleared = true
                clear(edgesSelectedTextLayer.context, Point.Zero, edgesSelectedTextLayer.getSize)
                edgesSelectedTextLayer.cleared = true
                clear(verticesSelectedLayer.context, Point.Zero, verticesSelectedLayer.getSize)
                verticesSelectedLayer.cleared = true
                clear(verticesSelectedTextLayer.context, Point.Zero, verticesSelectedTextLayer.getSize)
                verticesSelectedTextLayer.cleared = true

                draw(null, None, None)
                //^because elements are drawn into separate layers, redraw(..) does not know to which context to draw
            case RedrawOperation.Selection =>
                redrawAll()
            case _ =>
                redrawAll()
        }
    }

    def redrawAll() {
        layers.foreach {layer =>
            clear(layer.context, Point.Zero, layer.getSize)
            layer.cleared = true
        }
        draw(null, None, None)
        //^because elements are drawn into separate layers, redraw(..) does not know to which context to draw
    }
}
